File: src/com/themaskedcrusader/tmcz/modules/visibility/VisibilitySchedule.java
/*
 * Copyright 2013 Topher Donovan (themaskedcrusader.com)
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.themaskedcrusader.tmcz.modules.visibility;

import com.themaskedcrusader.bukkit.config.Settings;
import com.themaskedcrusader.bukkit.util.WorldUtils;
import com.themaskedcrusader.tmcz.data.PlayerUtil;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.HashMap;
import java.util.Map;

public class VisibilitySchedule {
    private static JavaPlugin plugin;
    static Map<String, Location> playerMovement = new HashMap<String, Location>();

    public VisibilitySchedule(JavaPlugin plugin) {
        VisibilitySchedule.plugin = plugin;

        plugin.getServer().getScheduler().scheduleSyncRepeatingTask(plugin, new Runnable() {
            public void run() {
                VisibilitySchedule.calculateVisibility();
            }
        }, 20L, Settings.getConfig().getLong(Visibility.TICKS));
    }

    public static void calculateVisibility() {
        Player[] players = plugin.getServer().getOnlinePlayers();
        for (Player player : players) {
            if (Visibility.isAllowed(player)) {
                float visibility = calculateVisibility(player);
                player.setExp(visibility / 7);
                playerMovement.put(player.getDisplayName(), player.getLocation());
            }
        }
    }

    private static float calculateVisibility(Player player) {
        float visibility = 2;
        Location loc = playerMovement.get(player.getDisplayName());
        if (((!Settings.getConfig().getBoolean(Visibility.IN_GAME)) || PlayerUtil.isPlaying(player)) && loc != null) {
            if (WorldUtils.isPlayerMoving(player, loc))  { visibility += 1;     }
            if (WorldUtils.isPlayerFalling(player, loc)) { visibility += 1;     }
            if (player.isSneaking())                     { visibility -= 1;     }
            if (player.isSprinting())                    { visibility += 2;     }
            if (player.getWorld().isThundering())        { visibility -= 0.5F;  }
            return visibility;
        }
        return 0f;
    }
}

File: src/main/kotlin/me/schalk/spigot/tmcz/modules/visibility/VisibilitySchedule.kt
/*
 * Copyright (c) 2013 - 2023 Christopher Schalk
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package me.schalk.spigot.tmcz.modules.visibility

import me.schalk.spigot.lib.config.getSettings
import me.schalk.spigot.lib.util.isPlayerFalling
import me.schalk.spigot.lib.util.isPlayerMoving
import me.schalk.spigot.tmcz.data.GameData
import org.bukkit.Location
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin

class VisibilitySchedule(plugin: JavaPlugin) : VisibilityModule() {

    init{
        VisibilitySchedule.plugin = plugin
        plugin.server.scheduler.scheduleSyncRepeatingTask(plugin,
            { calculateVisibility() }, 20L, getSettings().getConfig().getLong(MODULE + TICKS)
        )
    }

    companion object {
        private lateinit var plugin: JavaPlugin
        private var playerMovement = mutableMapOf<String, Location>()
        private const val _IN_GAME_ONLY = MODULE + IN_GAME_ONLY

        fun calculateVisibility() {
            val players = plugin.server.onlinePlayers
            players.forEach { player ->
                if (isAllowed(player)) {
                    val visibility = calculateVisibility(player)
                    player.exp = visibility / 7
                    playerMovement[player.displayName] = player.location
                }
            }
        }

        private fun calculateVisibility(player: Player): Float {
            var visibility = 2f
            val loc = playerMovement[player.displayName]
            if ((!getSettings().getConfig().getBoolean(_IN_GAME_ONLY) || GameData.getPlayer(player).playing) && loc != null) {
                if (isPlayerMoving(player, loc)) {
                    visibility += 1f
                }
                if (isPlayerFalling(player, loc)) {
                    visibility += 1f
                }
                if (player.isSneaking) {
                    visibility -= 1f
                }
                if (player.isSprinting) {
                    visibility += 2f
                }
                if (player.world.isThundering) {
                    visibility -= 0.5f
                }
                return visibility
            }
            return 0f
        }
    }

}